houston/specification.py

```diff
@@ -1,39 +1,341 @@
+        smt = Expression.values_to_smt('_', state.to_json(), decls)
+        smt.extend(self.get_expression(decls, state))
+        s.add(smt)
+        return s.check() == z3.sat
+
+    def get_expression(self,
+                       decls: Dict[str, Any],
+                       state: State,
+                       postfix: str = ""
+                       ) -> List[z3.ExprRef]:
+        """
+        Constructs a Z3 expression from this expression for a particular
+        state and set of declaration mappings.
+        """
+        expr = z3.parse_smt2_string('(assert {})'
+                                    .format(self.expression), decls=decls)
+        variables = {}
+        for v in state.variables:
+            variables['_{}{}'.format(v.name, postfix)] = float(v.noise) \
+                if v.is_noisy else 0.0
+            variables['__{}{}'.format(v.name, postfix)] = float(v.noise) \
+                if v.is_noisy else 0.0
+        expr_with_noise = [Expression.recreate_with_noise(expr, variables)]
+        return expr_with_noise
+
+    @staticmethod
+    def recreate_with_noise(expr: z3.ExprRef,
+                            variables: Dict[z3.ArithRef, float]
+                            ) -> z3.ExprRef:
+        recreate = Expression.recreate_with_noise
+        d = expr.decl()
+        if str(d) != '==':
+            children = [recreate(c, variables) for c in expr.children()]
+            # TODO find a better solution
+            if str(d) == 'And':
+                return z3.And(*children)
+            elif str(d) == 'Or':
+                return z3.Or(*children)
+            return d(*children)
+        lhs, rhs = expr.children()[0], expr.children()[1]
+        if not isinstance(lhs, z3.ArithRef) or \
+                not isinstance(rhs, z3.ArithRef):
+            return d(lhs, rhs)
+
+        def absolute(x):
+            return z3.If(x > 0, x, -x)
+        return absolute(lhs - rhs) <= Expression.get_noise(expr, variables)
+
+    @staticmethod
+    def get_noise(expr: z3.ExprRef,
+                  variables: Dict[z3.ArithRef, float]
+                  ) -> float:
+        if len(expr.children()) == 0:
+            if isinstance(expr, z3.ArithRef) and str(expr) in variables:
+                return variables[str(expr)]
+            else:
+                return 0.0
+        noises = []
+        for c in expr.children():
+            noises.append(Expression.get_noise(c, variables))
+        if str(expr.decl()) == '*':
+            f = 1.0
+            for i in noises:
+                f *= i
+            return f
+        elif str(expr.decl()) == '**':
+            if isinstance(expr.children()[1], z3.IntNumRef):
+                return math.pow(noises[0], int(expr.children()[1]))
+            else:
+                # TODO: FIX
+                return noises[0]
+        else:
+            return math.fsum(noises)
 
 
-@attr.s
 class Specification(object):
-    name = attr.ib(type=str)
```

Comment:
Use attrs (i.e., use the code that was deleted).

Reply:
I don't exactly know how to specify that the argument should be `str` but I want the value to be set as `Expression(precondition)` in this format. If you know how, let me know so that I can change it.